extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcIdentityProvider.java

```diff
@@ -276,18 +276,29 @@ private Uni<TokenVerificationResult> verifyCodeFlowAccessTokenUni(RoutingContext
                 && (resolvedContext.oidcConfig.authentication.verifyAccessToken
                         || resolvedContext.oidcConfig.roles.source.orElse(null) == Source.accesstoken)) {
             final String codeAccessToken = (String) vertxContext.get(OidcConstants.ACCESS_TOKEN_VALUE);
-            return verifyTokenUni(resolvedContext, codeAccessToken);
+            return verifyTokenUni(resolvedContext, codeAccessToken, null);
         } else {
             return NULL_CODE_ACCESS_TOKEN_UNI;
         }
     }
 
-    private Uni<TokenVerificationResult> verifyTokenUni(TenantConfigContext resolvedContext, String token) {
+    private Uni<TokenVerificationResult> verifyTokenUni(TenantConfigContext resolvedContext, String token, UserInfo userInfo) {
         if (OidcUtils.isOpaqueToken(token)) {
             if (!resolvedContext.oidcConfig.token.allowOpaqueTokenIntrospection) {
                 LOG.debug("Token is opaque but the opaque token introspection is not allowed");
                 throw new AuthenticationFailedException();
             }
+            // verify opaque access token with UserInfo if enabled and introspection URI is absent
+            if (resolvedContext.oidcConfig.token.verifyOpaqueAccessTokenWithUserInfo
+                    && resolvedContext.provider.getMetadata().getIntrospectionUri() == null) {
+                if (userInfo == null) {
+                    return Uni.createFrom().failure(
+                            new AuthenticationFailedException("Opaque access token verification failed as user info is null."));
+                } else {
+                    // valid token verification result
+                    return Uni.createFrom().item(new TokenVerificationResult(null, new TokenIntrospection("{}")));
```

Comment:
I'm not sure we need to emulate either the JWT verification result with the empty JSON or the introspection result which never happened - both values here should be null as we don't want the follow up code to work with these values, for example, `TokenIntrospection` injection may be requested and we don't want to meet that requirement in this case by injecting an empty bean. Having both of the values as null will obviously cause NPE or failure in the code in this class which creates `SecurityIdentity` so lets make sure that may be only an anonymous SecurityIdentity is created in this case but it will have a UserInfo attribute set for ex, etc, we need to see where these null values will be required and make sure no ambiguities are introduced via the empty JSONs

Reply:
I'll reflect comments later, but what you describe is by design and I thought it is desirable as current PR implementation:

- does not create any TokenIntrospection as it is a step above, `io.quarkus.oidc.runtime.OidcIdentityProvider#verifyTokenUni` is private method called in order to create identity
- if I send there null, than I have to add null check when security identity is created and that null value will be OK under certain circumstances - that adds complexity (not literally, you just want to limit null checks). 
- current implementation does only send response to internal  method that does not need to be modified - less complexity

Reply:
So we should be falling into this branch: https://github.com/quarkusio/quarkus/blob/48e0ea87ed85923a7d1b31b1e82906181cc700c5/extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcIdentityProvider.java#L194

But I think we need to isolate the code where a non-null introspection result is checked, this line, https://github.com/quarkusio/quarkus/blob/48e0ea87ed85923a7d1b31b1e82906181cc700c5/extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcIdentityProvider.java#L197,
will support `@Inject TokenIntrospection intr;` which should really fail in this case. So this line should be done only if the real introspection result is available, this is why we should return null earlier.
Please also tweak a bit the follow up principal check and have an empty principal name if the introspection is null. and it should be all

Reply:
I see your point, I completely missed that, thank you.